src/negate.c

```diff
@@ -0,0 +1,20 @@
+#include "data.table.h"
+
+void negateByRef(SEXP x) {
+  if(TYPEOF(x) != LGLSXP) {
+    error("not logical or integer vector");
+  }
+  int n = length(x);
```

Comment:
Good to declare this `const int n =`. That may help the compiler in the `for()` loop telling it that `n` can't change so it might be able to optimize `i<n` on every iteration (by doing it in batches or something clever like that). Since we target all sorts of compilers on many architectures, perhaps some of them benefit from the `const` declaration.